vignettes/creating-custom-modules.Rmd

```diff
@@ -15,170 +15,237 @@ vignette: >
 However, it is also possible to create your own modules using the `module` function.
 
-## Components of a module
+In this guide, we will use the simple histogram below as an example, and demonstrate how to convert this histogram function into a robust `teal` module step-by-step:
+```r
+my_plot <- hist(
+  dataset[[vars]],
+  las = 1,
+  main = paste("Histogram of", vars),
+  xlab = vars,
+  col = "lightblue",
+  border = "black"
+)
+```
+This module will allow users to dynamically select datasets and variables to create histograms within a `teal` application.
+We will cover best practices, including setting up dynamic inputs, structuring server logic, and using the `teal_data` object to ensure reactivity and reproducibility.
+
+## Understanding the Inputs and Requirements
+
+When developing a custom `teal` module for visualizations, we will first identify the primary inputs that users will interact with:
+
+* **Dataset Input** (`dataset`): Allows users to select which dataset to explore.
+* **Variable Input** (`selected`): Allows users to choose a specific numeric variable from the chosen dataset, ensuring only appropriate columns are available for plotting.
+
+These inputs are dynamically populated based on the available datasets and variables in the `teal_data` object, which we will cover later.
+
+## Setting Up the `teal` Module UI
+
+The UI function defines the controls and display area for the histogram. 
+For this module, we will use:
+
+- **`selectInput` for Dataset**: Enables users to select a dataset from the list of available datasets.
+- **`selectInput` for Variable**: Allows users to choose a numeric variable from the chosen dataset, dynamically filtering out any non-numeric columns.
+- **`plotOutput` for Histogram**: Displays the histogram once both dataset and variable inputs are selected.
+- **`verbatimTextOutput` for Code**: Automatically displays code that generated the plot based on user input.
+
+Here’s the code for the `histogram_module_ui` function:
+
+```r
+# UI function for the custom histogram module
+histogram_module_ui <- function(id) {
+  ns <- shiny::NS(id)
+  shiny::tagList(
+    shiny::selectInput(ns("dataset"), "Select Dataset", choices = NULL),
+    shiny::selectInput(ns("variable"), "Select Variable", choices = NULL),
+    shiny::plotOutput(ns("histogram_plot")),
+    shiny::verbatimTextOutput(ns("plot_code"))  # To display the reactive plot code
+  )
+}
+```
+
+## Setting Up the `teal` Module Server
 
-### UI function
+The server function is where the main logic of a `teal` module is handled. 
+For our histogram module, the server function will handle user interactions and manage the reactive `teal_data` object, which allows the module to dynamically respond to user inputs.
 
-This function contains the UI required for the module.
-It should be a function with at least the argument `id`.
-See the server section below for more details.
+### Passing the `data` Argument to the Server Function
 
```

Comment:
I'm a bit on the fence about this because I’m hoping users have already reviewed the "Getting Started with teal" vignette, which covers `teal_data`.
https://insightsengineering.github.io/teal/latest-tag/articles/getting-started-with-teal.html#application-data

<br />
<br />

> Link that was in the previous documentation https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data-reproducibility.html

The link provided in the previous documentation made sense because the sentence before it discussed code reproducibility. Including it here without any lead-in context doesn’t quite fit.

If we think it’s needed, we could close this sub-category with a reminder sentence like this:

_If you need a refresher on the teal_data object, please visit the [teal.data package documentation](https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data.html)._

What do you think?

Reply:
> If you need a refresher on the teal_data object, please visit the [teal.data package documentation](https://insightsengineering.github.io/teal.data/latest-tag/articles/teal-data.html).

:+1:

Reply:
Done with 5617508983db5686177b3260a469fdbb71a0e6fb